nixos/boot: rename network interfaces already in stage 1

xeji opened the pull request:
###### Motivation for this change

Fix #39069. When `networking.usePredictableInterfaceNames = true`, rename network interfaces already in stage 1 boot  to avoid a race condition between interface renaming and configuration.
This also ensures consistent interface names between stage 1 (initrd) and fully booted system.

~~~Please backport to 18.03 because a resulting non-deterministic test failure has frequently delayed the `nixos-18.03-small` channel.~~~ (better not, see discussion)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

xeji commented:
Yes, this might break initrd networking if you have custom stage 1 code containing hard coded references to `eth0` etc. 

Definitely an issue for 18.03. But I think we can make this change on master even without `configVersion`. Unstable means that things may break, and for 18.09 we would document it in the release notes.

There's also an alternative (ugly) solution that leaves stage 1 unchanged: issue an extra `udevadm trigger -s; udevadm settle` restricted to net subsystem at the start of stage 2 before starting `systemd`. Would work but we don't really want to call udev 3 times in 2 boot stages...

oxij commented:
> Yes, this might break initrd networking if you have custom stage 1 code containing hard coded references to `eth0` etc. 

It could break it even without anything custom because nix has no way of knowing which interfaces exist on a booted system, hence it can't check/warn about anything. Including "ip" kernel parameter `initrd.network` asks you to use.

> breaking master

Well, I'm ok with breaking master if no trivial change can fix it, but `configVersion` is a fairly trivial patchset and using it here would fix everything.

xeji commented:
Ok, so let's first mplement the ugly alternative I mentioned (it's better for 18.03 anyway) and save this for later. 

We should definitely get this done in 18.09. Having interface names change between initrd and full boot doesn't make sense.

xeji commented:
Non-breaking interim fix in #39340. Marked this as [WIP] until we can get this into 18.09.

globin commented:
@xeji any updates on this?

xeji commented:
I would like to get this ready now - shouldn't take long, mainly adapt to systemd 239 and test.

Before proceeding we still need to address concerns about breaking existing configs, see discussion above. In the worst case, servers using initrd networking could become unreachable after update. 

I see two options: 

1. accept this breakage and document it in the release notes
2. make this change conditional on `stateVersion` - safer but leads to uglier code. (@oxij mentioned a `configVersion` mechanism but I don't think that has been implemented yet)

How should be proceed here?

xeji commented:
Made the change conditional on `stateVersion` and added documentation. First tests look ok, but needs more testing.

xeji commented:
We could go a step further and dump our old non-standard `80-net-setup-link.rules` in favor of the standard systemd udev rule, which uses an udev builtin to rename interfaces to predictable names based on the `net.ifnames` kernel parameter. Then `usePredictableInterfaceNames` becomes essentially an alias for that kernel parameter. But this will make it difficult to support the previous behavior for older configs, so I think we should do that maybe in 19.03 when everyone has had a chance to update their configs.

xeji committed:
predictable interface names: release notes entry

for changed behavior when stateVersion >= 18.09

xeji committed:
nixos/tests/predicatble-interface-names: stateVersion = 18.09

- test new implementation with system.stateVersion = 18.09
- enable initrd networking to see interface names in logs

xeji committed:
nixos/networkd: don't wait for systemd-udev-settle

not necessary if stateVersion >= 18.09
as devices are already renamed in stage 1

xeji committed:
networking.usePredictableInterfaceNames: document new behavior

xeji committed:
boot/stage-1: rename network interfaces in stage 1

if networking.usePredictableInterfaceNames=true
and system.stateVersion >= 18.09

xeji committed:
nixos/tests/initrd-network-ssh: also test predictable interface names

because now interfaces are already renamed in stage 1

xeji commented:
Refactored and added a test for initrd networking (which is the case breaking old configs).
This is ready for review now.

samueldr reviewed (commented):
🎉 Let's try to sneak this in before the branch-off.

It seems generally good, have a couple nits in the documentation, but I won't hang the review if the changes aren't exactly as asked.

samueldr commented:
@GrahamcOfBorg test initrd-network-ssh

GrahamcOfBorg commented:
Success on aarch64-linux

Attempted: tests.initrd-network-ssh

No partial log is available.

GrahamcOfBorg commented:
Success on x86_64-linux

Attempted: tests.initrd-network-ssh

No partial log is available.

xeji commented on nixos/modules/services/hardware/udev.nix:
```diff
@@ -266,6 +266,9 @@ in
         unpredictably.
+        For newer configurations with <literal>system.stateVersion >= 18.09</literal>,
+        interfaces are renamed to predictable names in Stage 1 (initrd), and
+        changing this option's value requires a reboot to take effect.
```
> Reading this, I initially thought that this was related to the previous sentence. 

It actually is related - in the current implementation, the option can be changed with `nixos-rebuild switch` without rebooting. But with this PR it will require a reboot.

xeji commented:
@samueldr you asked in a comment if we should also test stateVersion=18.03. When I tried to post my answer, my browser got stuck, and after reload your comment and my answer were both gone. Don't know what happened....

Anyway, we can also add the tests for stateVersion=18.03. I was a bit reluctant to add that variable because that would increase it from 4 to 8 tests. But I can add them to be more complete.

samueldr commented:
Sorry for your browser, I removed the comment since I thought it was irrelevant; I hadn't read thoroughly the initrd SSH test and thought they were linked.

If it turns out to be relevant, maybe at first building the to allow multiple stateVersion, and in the future when deemed appropriate it can be disabled. Having the plumbing will allow future changes requiring stateVersion to be more easily tested, hopefully.

xeji committed:
predictable interface names: shorten release notes entry

xeji committed:
networking.usePredictableInterfaceNames: document new behavior (update)

xeji commented:
Updated documentation. Will clean up the commits before merge.

xeji committed:
nixos/tests/predictableInterfaceNames: test stateVersion 18.03 and 18.09

GrahamcOfBorg commented:
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39329&attempt_id=d531a0c8-6138-477c-9830-d15fcca9d901)

Attempted: tests.initrd-network-ssh.unpredictable, tests.initrd-network-ssh.predictable

<details><summary>Partial log (click to expand)</summary><p>

```
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
building of '/nix/store/valydyj9spbbfw56k77l3blv6d1y0cv6-vm-test-run-initrd-network-sshpredictable.drv' timed out after 3600 seconds
error: build of '/nix/store/valydyj9spbbfw56k77l3blv6d1y0cv6-vm-test-run-initrd-network-sshpredictable.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Timed out, unknown build status on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39329&attempt_id=aea08644-8803-4204-81bd-bf6193666fd3)

Attempted: tests.initrd-network-ssh.unpredictable, tests.initrd-network-ssh.predictable

<details><summary>Partial log (click to expand)</summary><p>

```
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
building of '/nix/store/6qv2bhzr2jr9lqv9gapp9ziamjq72jvx-vm-test-run-initrd-network-sshpredictable.drv' timed out after 3600 seconds
error: build of '/nix/store/6qv2bhzr2jr9lqv9gapp9ziamjq72jvx-vm-test-run-initrd-network-sshpredictable.drv' failed
```
</p></details>

xeji commented:
Re timeout: Looks like this failed because ofborg started both test variants in parallel as one build job because they were given in one command, and they interfered with each other. Seen this before, I think you need to trigger tests separately on ofborg:

@GrahamcOfBorg test initrd-network-ssh.predictable
@GrahamcOfBorg test initrd-network-ssh.unpredictable

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39329&attempt_id=7f41085c-82ca-4e08-afb7-4c52e22867d5)

Attempted: tests.initrd-network-ssh.unpredictable

<details><summary>Partial log (click to expand)</summary><p>

```
client# [   26.277303] reboot: Power down
vde_switch: EOF data port: Interrupted system call
collecting coverage data
syncing
test script finished in 27.20s
cleaning up
killing server (pid 653)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/ihs0q9mvmv0abjm6vg91ryvpsf8mx1gf-vm-test-run-initrd-network-ssh
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39329&attempt_id=ceff06ad-9d51-4dd9-89cf-d8724b235dcd)

Attempted: tests.initrd-network-ssh.unpredictable

<details><summary>Partial log (click to expand)</summary><p>

```
client# [   25.200972] reboot: Power down
vde_switch: EOF data port: Interrupted system call
collecting coverage data
syncing
test script finished in 26.74s
cleaning up
killing server (pid 622)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/ajacg9dk1lshv96b4yjc2ffa98w96q38-vm-test-run-initrd-network-ssh
```
</p></details>

samueldr commented:
Oh, though I did run the tests locally successfully using `-A .. -A ..` locally and it worked fine ¯\\\_(ツ)\_/¯

xeji commented:
Yeah, ofborg does that in a strange way. Has fooled me before.

xeji commented:
@GrahamcOfBorg test predictableInterfaceNames.predictable1809

xeji commented:
The "predictable" tests fail on aarch64 because the tests expect hardcoded `ens3` and `ens4`, and the predictable names on arm are obviously different. That's not new (was the same without this PR): 
https://hydra.nixos.org/job/nixos/unstable-aarch64/nixos.tests.predictable-interface-names.vm-test-run-predictableInterfaceNames.aarch64-linux/all never succeeded on Hydra.

GrahamcOfBorg commented:
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39329&attempt_id=4ba85492-d21c-408e-8293-7138f12449ca)

Attempted: tests.initrd-network-ssh.predictable

<details><summary>Partial log (click to expand)</summary><p>

```
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
building of '/nix/store/92cijh2iqx8c2ysmfpaqhaiw8j7zncfg-vm-test-run-initrd-network-sshpredictable.drv' timed out after 1800 seconds
error: build of '/nix/store/92cijh2iqx8c2ysmfpaqhaiw8j7zncfg-vm-test-run-initrd-network-sshpredictable.drv' failed
```
</p></details>

xeji commented:
Well, I see what happened here: on the ofborg builder the relevant interface has the name `ens8`. On my machine (and in the tests) it's `ens4`. With initrd networking you need to give `udhcpc` the interface name as an argument (unless it's the default `eth0`), so this failed to configure the interface. 

This is strange. I would have thought that our qemu test setup would give VMs the same emulated hardware on every host, so interfaces should always have the same predictable names. Apparently not.

Don't really know how to fix this. If we cannot rely on qemu giving a reproducible environment on each build host, we're fundamentally screwed...

xeji commented:
On a positive note: the predictable-interface-names tests succeed on Hydra, so this may be an ofborg-specific problem and shouldn't keep us from merging this.

GrahamcOfBorg commented:
Timed out, unknown build status on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39329&attempt_id=d85315f9-a621-4f71-9542-c4bb9b384c2a)

Attempted: tests.initrd-network-ssh.predictable

<details><summary>Partial log (click to expand)</summary><p>

```
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
building of '/nix/store/081zagf5asipv62a26vwjylhjykzr4wz-vm-test-run-initrd-network-sshpredictable.drv' timed out after 3600 seconds
error: build of '/nix/store/081zagf5asipv62a26vwjylhjykzr4wz-vm-test-run-initrd-network-sshpredictable.drv' failed
```
</p></details>

samueldr commented:
Running on the aarch64 community box I am seeing issues, and they seem consistent.

<details><summary>Partial output of <tt>nix-build nixos/tests/initrd-network-ssh/default.nix -A predictable</tt></summary>
```
server#
server# <<< NixOS Stage 1 >>>
server#
server# loading module virtio_balloon...
server# modprobe: FATAL: Module virtio_balloon not found in directory /lib/modules/4.14.66
server# loading module virtio_console...
server# modprobe: FATAL: Module virtio_console not found in directory /lib/modules/4.14.66
server# loading module virtio_rng...
server# [    1.824348] random: fast init done
server# [    1.828758] random: crng init done
server# loading module dm_mod...
server# [    1.850868] device-mapper: ioctl: 4.37.0-ioctl (2017-09-20) initialised: dm-devel@redhat.com
server# loading module af_packet...
server# modprobe: FATAL: Module af_packet not found in directory /lib/modules/4.14.66
server# running udev...
server# [    1.886406] systemd-udevd[141]: starting version 239
server# [    1.898005] usb 1-1: new high-speed USB device number 2 using ehci-pci
server# [    2.033234] virtio_net virtio0 enp0s1: renamed from eth0
server# [    2.075432] input: QEMU QEMU USB Keyboard as /devices/platform/3f000000.pcie/pci0000:00/0000:00:08.0/usb1/1-1/1-1:1.0/0003:0627:0001.0001/input/input1
server# [    2.086893] virtio_net virtio1 enp0s2: renamed from eth1
server# [    2.162322] hid-generic 0003:0627:0001.0001: input: USB HID v1.11 Keyboard [QEMU QEMU USB Keyboard] on usb-0000:00:08.0-1/input0
server# kbd_mode: KDSKBMODE: Inappropriate ioctl for device
server# %Gipconfig: ens4: SIOCGIFINDEX: No such device
server# ipconfig: no devices to configure
server# bringing up network interface enp0s1...
server# bringing up network interface enp0s2...
server# bringing up network interface lo...
server# [    2.306066] usb 1-2: new high-speed USB device number 3 using ehci-pci
server# acquiring IP address via DHCP...
server# udhcpc: SIOCGIFINDEX: No such device
server# [    2.473022] input: QEMU QEMU USB Tablet as /devices/platform/3f000000.pcie/pci0000:00/0000:00:08.0/usb1/1-2/1-2:1.0/0003:0627:0001.0002/input/input2
server# [    2.489920] hid-generic 0003:0627:0001.0002: input: USB HID v0.01 Mouse [QEMU QEMU USB Tablet] on usb-0000:00:08.0-2/input0
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
client: exit status 1
client: running command: ping -c 1 server
error: interrupted by the user

```
</details>

`-A unpredictable` is fine though.

xeji commented:
I'm speculating, but this may be caused by older versions of NixOS/nix/qemu on different builders. 

I will try to make the `predictable-interface-names` test more robust by grepping for a pattern instead of a hardcoded interface name. That should work around this issue and fix it on aarch64.

The `initrd-networking-ssh.predictable` test cannot be fixed this way because the expression must hardcode the interface name to start initrd networking. It will always fail on builders where that name is different. But that's a separate issue regarding uniform setup of our CI infrastructure and/or reproducible results of our qemu-based test framework - out of scope of this PR.

edolstra commented:
This should *not* depend on `stateVersion` because that's not what `stateVersion` is for. `stateVersion` is for cases where there is mutable state on disk (like PostgreSQL databases) that cannot be migrated easily. It does not mean "change option defaults to whatever was the case in NixOS version <X>".

xeji commented:
@edolstra then what mechanism do you suggest for preventing breakage of existing configs?

samueldr commented:
@edolstra other than `stateVersion` what can be used **now** to prevent breakage from this change?

Though, maybe in the end there are no risks? Would the system collect the currently booted configuration? There should be references still existing keeping it alive, right?

xeji commented:
This isn't ready yet, the tests still needs a little fine tuning (see above discussion). I suspended working on it until there's consensus on `stateVersion` or another approach.

mmahut commented:
Any updates on this pull request, please?

emilazy commented:
The `nixos/modules/services/hardware/80-net-setup-link.rules` file that this copies to initrd has been deleted as of 1f03f6f; what's the correct way forward to have this behaviour in initrd now that it has apparently been obsoleted? The commit message claims that the .rules files are now obsolete and all the renaming is done in C in udev, but my initrd seems to still use `eth0` and then the device renaming fails after boot:

```
Sep 16 14:04:39 patchouli systemd-udevd[1680]: eth0: Failed to rename network interface 2 from 'eth0' to 'enp4s0': Device or resource busy
```

cc @xeji @andir (author of the commit) @erikarvstedt (author of similar PR #47664)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/predictable-network-interface-names-in-initrd/4055/1

flokli commented:
I think this fix is included in https://github.com/NixOS/nixpkgs/pull/79532. @xeji, @oxij can you take a look?

flokli commented:
No reaction after https://github.com/NixOS/nixpkgs/pull/39329#issuecomment-604907307, and this is probably fixed, so closing.
